dagql/idtui/frontend.go

```diff
@@ -17,631 +16,75 @@ import (
-	defer func() {
-		if r := recover(); r != nil {
-			fe.restore()
-			panic(r)
-		}
-	}()
-	return doneMsg{fe.run(fe.runCtx)}
 }
 
-type backgroundDoneMsg struct{}
-
-func (fe *Frontend) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
-	switch msg := msg.(type) {
-	case doneMsg: // run finished
-		slog.Debug("run finished", "err", msg.err)
-		fe.done = true
-		fe.err = msg.err
-		if fe.eof {
-			return fe, tea.Quit
-		}
-		return fe, nil
-
-	case eofMsg: // received end of updates
-		slog.Debug("got EOF")
-		fe.eof = true
-		if fe.done {
-			return fe, tea.Quit
-		}
-		return fe, nil
-
-	case backgroundMsg:
-		fe.backgrounded = true
-		return fe, tea.Exec(msg.cmd, func(err error) tea.Msg {
-			msg.errs <- err
-			return backgroundDoneMsg{}
-		})
-
-	case backgroundDoneMsg:
-		return fe, nil
-
-	case tea.KeyMsg:
-		switch msg.String() {
-		case "q", "esc", "ctrl+c":
-			if fe.interrupted {
-				slog.Warn("exiting immediately")
-				return fe, tea.Quit
-			} else {
-				slog.Warn("canceling... (press again to exit immediately)")
-			}
-			fe.interrupt()
-			fe.interrupted = true
-			return fe, nil // tea.Quit is deferred until we receive doneMsg
-		case "ctrl+\\": // SIGQUIT
-			fe.restore()
-			sigquit()
-			return fe, nil
-		default:
-			return fe, nil
-		}
-
-	case tea.WindowSizeMsg:
-		fe.SetWindowSize(msg)
-		return fe, nil
+type Frontend interface {
+	// Run starts a frontend, and runs the target function.
+	Run(ctx context.Context, opts FrontendOpts, f func(context.Context) error) error
 
-	case ui.FrameMsg:
-		fe.render()
-		// NB: take care not to forward Frame downstream, since that will result
-		// in runaway ticks. instead inner components should send a SetFpsMsg to
-		// adjust the outermost layer.
-		return fe, ui.Frame(fe.fps)
-
-	default:
-		return fe, nil
-	}
-}
-
-func (fe *Frontend) SetWindowSize(msg tea.WindowSizeMsg) {
-	fe.window = msg
-	fe.db.SetWidth(msg.Width)
-	fe.messagesView.SetWidth(msg.Width)
-}
+	// SetPrimary tells the frontend which span should be treated like the focal
+	// point of the command. Its output will be displayed at the end, and its
+	// children will be promoted to the "top-level" of the TUI.
+	SetPrimary(spanID trace.SpanID)
+	Background(cmd tea.ExecCommand) error
 
-func (fe *Frontend) render() {
-	fe.mu.Lock()
-	fe.view.Reset()
-	fe.Render(NewOutput(fe.view, termenv.WithProfile(fe.profile)))
-	fe.mu.Unlock()
-}
+	// Can consume otel spans and logs.
+	sdktrace.SpanExporter
+	sdklog.LogExporter
 
-func (fe *Frontend) View() string {
-	fe.mu.Lock()
-	defer fe.mu.Unlock()
-	view := fe.view.String()
-	if fe.backgrounded {
-		// if we've been backgrounded, show nothing, so a user's shell session
-		// doesn't have any garbage before/after
-		return ""
-	}
-	if fe.done && fe.eof {
-		// print nothing; make way for the pristine output in the final render
-		return ""
-	}
-	return view
+	// ConnectedToEngine is called when the CLI connects to an engine.
+	// TODO: remove this
```

Comment:
Mostly curious, any particular reason to remove it? The thinking was that a plain UI would just print this, but a pretty TUI might show it somewhere onscreen

Reply:
It gets refactored away in https://github.com/dagger/dagger/pull/7385/files#diff-375c41f6a94e797edee687b8fa151865e2192d8aa9088acaffa174d0de2c3578R284

We can just use `slog` here (at least for now).